server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java

```diff
@@ -20,30 +20,77 @@
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.transport.TransportAddress;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.ObjectParser;
 import org.elasticsearch.common.xcontent.ToXContentFragment;
 import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
 import org.elasticsearch.node.Node;
 
 import java.io.IOException;
+import java.net.InetAddress;
 import java.util.Collections;
 import java.util.EnumSet;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Set;
 import java.util.function.Predicate;
 
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+
 
 /**
  * A discovery node represents a node that is part of the cluster.
  */
 public class DiscoveryNode implements Writeable, ToXContentFragment {
 
+    private static final ParseField NAME = new ParseField("name");
+    private static final ParseField EPHEMERAL_ID = new ParseField("ephemeral_id");
+    private static final ParseField TRANSPORT_ADDRESS = new ParseField("transport_address");
+    private static final ParseField ATTRIBUTES = new ParseField("attributes");
+
+    private static final ConstructingObjectParser<DiscoveryNode, String> PARSER =
+        new ConstructingObjectParser<>("innerDiscoveryNode",
+        true,
+            (a, c) -> {
+                final String nodeName = (String) a[0];
+                final String nodeId = c;
+                final String ephemeralId = (String) a[1];
+                final TransportAddress address = (TransportAddress) a[2];
+                final String hostName = address.address().getHostString();
+                final String hostAddress = address.getAddress();
+                final Map<String, String> attributes = (Map<String, String>) a[3];
+                // TODO: impossible to recreate roles
+                final Set<Role> roles = Collections.emptySet();
+                // TODO: impossible to recreate version
+                final Version version = Version.CURRENT;
+
+                return new DiscoveryNode(nodeName, nodeId, ephemeralId, hostName, hostAddress, address, attributes,
+                    roles, version);
+            });
+
+    static {
+        PARSER.declareField(constructorArg(), (p, c) -> p.text(), NAME, ObjectParser.ValueType.STRING);
+        PARSER.declareField(constructorArg(), (p, c) -> p.text(), EPHEMERAL_ID, ObjectParser.ValueType.STRING);
+        PARSER.declareField(constructorArg(), (p, c) -> {
+            final String text = p.text();
+            final int colonIdx = text.indexOf(':');
+            final String host = text.substring(0, colonIdx);
+            final int port = Integer.parseInt(text.substring(colonIdx + 1));
+            // TODO: ACHTUNG ! DNS lookup !
```

Comment:
Yes :( :( :( :( this is why I did not want to parse DiscoveryNode, I think we may want to have a intermediate object for representation here. I have an idea for this we can discuss